File: ddpoliglot_v6_flutter/lib/models/data/WordTranslation.dart
// ignore_for_file: file_names, hash_and_equals, annotate_overrides
import '../player_classes.dart';
import 'Word.dart';
import './Language.dart';

import 'package:json_annotation/json_annotation.dart';

part 'WordTranslation.g.dart';

@JsonSerializable()
class WordTranslation extends Plailable {
  int wordTranslationID;
  int wordID;
  Word? word;
  int languageID;
  Language? language;
  String text;
  int hashCode;
  String? textSpeechFileName;
  int speachDuration;
  int readyForLessonPhrasiesCnt; // phrases for this word, ordered, translated and speeched for this language

  WordTranslation(
    this.wordTranslationID,
    this.wordID,
    this.word,
    this.languageID,
    this.language,
    this.text,
    this.hashCode,
    this.textSpeechFileName,
    this.speachDuration,
    this.readyForLessonPhrasiesCnt, // phrases for this word, ordered, translated and speeched for this language
  );

  factory WordTranslation.fromJson(Map<String, dynamic> data) =>
      _$WordTranslationFromJson(data);

  Map<String, dynamic> toJson() => _$WordTranslationToJson(this);
}

File: ddPoliglotV6.Data/Models/WordTranslation.cs
using System;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Text;

namespace ddPoliglotV6.Data.Models
{
    public class WordTranslation
    {
        [Key]
        public int WordTranslationID { get; set; }

        public int WordID { get; set; }

        [ForeignKey("WordID")]
        public Word Word { get; set; }

        public int LanguageID { get; set; }

        [ForeignKey("LanguageID")]
        public Language Language { get; set; }

        [MaxLength(250)]
        public string Text { get; set; }
        public Int64 HashCode { get; set; }

        [MaxLength(100)]
        public string TextSpeechFileName { get; set; }

        public int SpeachDuration { get; set; }

        public int ReadyForLessonPhrasiesCnt { get; set; } // phrases for this word, ordered, translated and speeched for this language

        public object Clone()
        {
            return this.MemberwiseClone();
        }

        public string Compare(WordTranslation other)
        {
            string[] propNames = new string[] {
                "LanguageID",
                "Text",
                "HashCode",
                "TextSpeechFileName",
                "SpeachDuration",
                "ReadyForLessonPhrasiesCnt",
            };

            var result = new StringBuilder();
            foreach (var propName in propNames)
            {
                CompareProp(this, other, propName, result);
            }

            return result.ToString();
        }

        private void CompareProp(object curr, object other, string propName, StringBuilder result)
        {
            var currValue = curr.GetType().GetProperty(propName).GetValue(curr, null);
            var otherValue = other.GetType().GetProperty(propName).GetValue(other, null);
            if ((currValue?.ToString() ?? "null") != (otherValue?.ToString() ?? "null"))
            {
                result.AppendLine($"{propName}: {currValue} -> {otherValue}");
            }
        }
    }
}
